### ISSUE TITLE
Normalisation error in torch_geometric.utils.softmax() with extreme values.

### ISSUE THREAD

**Issue by AndrewPaulChester** [rusty1s/pytorch_geometric#2274]
## 🐛 Bug

When extreme values are present in the src tensor for torch_geometric.utils.softmax(), it can affect the normalisation of other groups (i.e. with different indexes)

## To Reproduce

Here is a minimal working example:

```
import torch
from torch_geometric.utils import softmax

broken = torch.tensor([1, -1, -np.inf, 50])
index = torch.tensor([1, 1, 2, 2])
print(softmax(broken[0:2],index[0:2]))
print(softmax(broken[0:4],index[0:4]))

working = torch.tensor([1, -1, -np.inf, 30])
print(softmax(working[0:2],index[0:2]))
print(softmax(working[0:4],index[0:4]))

```
Which produces output:

```
tensor([0.8808, 0.1192])
tensor([5.2429e-06, 7.0954e-07, 0.0000e+00, 1.0000e+00])
tensor([0.8808, 0.1192])
tensor([0.8805, 0.1192, 0.0000, 1.0000])

```

## Expected behavior

Expected behaviour is that adding new groups into the softmax function should not affect other groups, i.e. the outputs for broken and working tensors should be identical. As far as I can tell the ratio of the affected values is still correct, they are just not normalised to sum to 1. 

## Environment

* OS: Ubtuntu (checked both 20.04 under WSL2 and 16.04 natively)
* Python version: 3.8.5
* PyTorch version: 1.7.1
* CUDA/cuDNN version: 11.3
* GCC version: 9.3.0
* Any other relevant information:
* PyTorch Geometric version: 1.6.3

## Additional context
I encountered this while debugging NaN values that were popping up during backprop. I am not 100% sure of this, but I think that this is the cause, because the gradients are getting mangled somehow during this error. Given that this is happening when it's normalising over -np.inf and some large positive value, I'm guessing it might be over/underflowing somewhere.

**Comment by ldv1** [comment]
With master and PyTorch 1.8 I get

```
tensor([0.8808, 0.1192])
tensor([0.8808, 0.1192, 0.0000, 1.0000])
tensor([0.8808, 0.1192])
tensor([0.8808, 0.1192, 0.0000, 1.0000])

```

**Comment by rusty1s** [comment]
Indeed, this is already fixed in PyG master :)
